application/common/components/Emailer.php

```diff
@@ -410,24 +421,66 @@ public function sendDelayedMfaRelatedEmails()
     }
 
     /**
-     *
      * Whether the user has already been sent this type of email in the last X days
      *
      * @param int $userId
      * @param string $messageType
      * @return bool
      */
-    public function hasReceivedMessageRecently($userId, string $messageType)
+    public function hasUserReceivedMessageRecently($userId, string $messageType): bool
     {
-        $latestEmail = EmailLog::find()->where(['user_id' => $userId, 'message_type' => $messageType])
-            ->orderBy('sent_utc DESC')->one();
+        $user = User::findOne($userId);
+        if (empty($user)) {
+            \Yii::warning(sprintf(
+                'User (%s) not found when checking for recent emails',
+                $userId
+            ));
+            return false;
+        }
+        return $this->hasAddressReceivedMessageRecently($user->email, $messageType);
+    }
+
+    public function hasAddressReceivedMessageRecently(?string $toAddress, string $messageType): bool
```

Comment:
You can avoid the `?` on the parameter type. The email address is never null, I believe.
```suggestion
    public function hasAddressReceivedMessageRecently(string $toAddress, string $messageType): bool
```

Reply:
I think (as you say) we are not ever currently passing in anything other than a string. However, depending on that seems a little fragile. For example, the `hrNotificationsEmail` value is initially null. We have a check in `Emailer->init()` to see if it is, and if so change it to an empty string, but there's nothing about the property that guarantees that it's always a string.

It seems more resilient to allow this to be null, and simply handle it internally (as the `empty` check does), if you don't mind me leaving it this way.

Reply:
That's fine, thanks.

Reply:
debugging parameter type issues is painful!